File: sdk-rust/src/models/vault.rs
use serde::{Deserialize, Serialize};

use super::common::VaultType;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Vault {
    pub address: String,
    pub name: String,
    pub description: Option<String>,
    pub manager: String,
    pub status: String,
    pub created_at: i64,
    pub tvl: Option<f64>,
    pub volume: Option<f64>,
    pub volume_30d: Option<f64>,
    pub all_time_pnl: Option<f64>,
    pub net_deposits: Option<f64>,
    pub all_time_return: Option<f64>,
    pub past_month_return: Option<f64>,
    pub sharpe_ratio: Option<f64>,
    pub max_drawdown: Option<f64>,
    pub weekly_win_rate_12w: Option<f64>,
    pub profit_share: Option<f64>,
    pub pnl_90d: Option<f64>,
    pub manager_cash_pct: Option<f64>,
    pub average_leverage: Option<f64>,
    pub depositors: Option<i64>,
    pub perp_equity: Option<f64>,
    pub vault_type: Option<VaultType>,
    pub social_links: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VaultsResponse {
    pub items: Vec<Vault>,
    pub total_count: i64,
    pub total_value_locked: f64,
    pub total_volume: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserOwnedVault {
    pub vault_address: String,
    pub vault_name: String,
    pub vault_share_symbol: String,
    pub status: String,
    pub age_days: i64,
    pub num_managers: i64,
    pub tvl: Option<f64>,
    pub apr: Option<f64>,
    pub manager_equity: Option<f64>,
    pub manager_stake: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VaultDeposit {
    pub amount_usdc: f64,
    pub shares_received: f64,
    pub timestamp_ms: i64,
    pub unlock_timestamp_ms: Option<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VaultWithdrawal {
    pub amount_usdc: Option<f64>,
    pub shares_redeemed: f64,
    pub timestamp_ms: i64,
    pub status: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserPerformanceOnVault {
    pub vault: Vault,
    pub account_address: String,
    pub total_deposited: Option<f64>,
    pub total_withdrawn: Option<f64>,
    pub current_num_shares: Option<f64>,
    pub current_value_of_shares: Option<f64>,
    pub share_price: Option<f64>,
    pub all_time_earned: Option<f64>,
    pub all_time_return: Option<f64>,
    pub volume: Option<f64>,
    pub weekly_win_rate_12w: Option<f64>,
    pub deposits: Option<Vec<VaultDeposit>>,
    pub withdrawals: Option<Vec<VaultWithdrawal>>,
    pub locked_amount: Option<f64>,
    pub unrealized_pnl: Option<f64>,
}

// --- Vault Operation Args ---

#[derive(Debug, Clone)]
pub struct CreateVaultArgs {
    pub vault_name: String,
    pub vault_description: String,
    pub vault_social_links: Vec<String>,
    pub vault_share_symbol: String,
    pub vault_share_icon_uri: Option<String>,
    pub vault_share_project_uri: Option<String>,
    pub fee_bps: u64,
    pub fee_interval_s: u64,
    pub contribution_lockup_duration_s: u64,
    pub initial_funding: u64,
    pub accepts_contributions: bool,
    pub delegate_to_creator: bool,
    pub contribution_asset_type: Option<String>,
    pub subaccount_addr: Option<String>,
}

#[derive(Debug, Clone)]
pub struct ActivateVaultArgs {
    pub vault_address: String,
    pub additional_funding: Option<u64>,
}

#[derive(Debug, Clone)]
pub struct DepositToVaultArgs {
    pub vault_address: String,
    pub amount: u64,
}

#[derive(Debug, Clone)]
pub struct WithdrawFromVaultArgs {
    pub vault_address: String,
    pub shares: u64,
}

File: sdk-python/decibel_sdk/models/vault.py
"""Vault models and operation argument types."""

from __future__ import annotations

from dataclasses import dataclass, field

from pydantic import BaseModel

from decibel_sdk.models.common import VaultType


class Vault(BaseModel):
    address: str
    name: str
    description: str | None = None
    manager: str
    status: str
    created_at: int
    tvl: float | None = None
    volume: float | None = None
    volume_30d: float | None = None
    all_time_pnl: float | None = None
    net_deposits: float | None = None
    all_time_return: float | None = None
    past_month_return: float | None = None
    sharpe_ratio: float | None = None
    max_drawdown: float | None = None
    weekly_win_rate_12w: float | None = None
    profit_share: float | None = None
    pnl_90d: float | None = None
    manager_cash_pct: float | None = None
    average_leverage: float | None = None
    depositors: int | None = None
    perp_equity: float | None = None
    vault_type: VaultType | None = None
    social_links: list[str] | None = None


class VaultsResponse(BaseModel):
    items: list[Vault]
    total_count: int
    total_value_locked: float
    total_volume: float


class UserOwnedVault(BaseModel):
    vault_address: str
    vault_name: str
    vault_share_symbol: str
    status: str
    age_days: int
    num_managers: int
    tvl: float | None = None
    apr: float | None = None
    manager_equity: float | None = None
    manager_stake: float | None = None


class VaultDeposit(BaseModel):
    amount_usdc: float
    shares_received: float
    timestamp_ms: int
    unlock_timestamp_ms: int | None = None


class VaultWithdrawal(BaseModel):
    amount_usdc: float | None = None
    shares_redeemed: float
    timestamp_ms: int
    status: str


class UserPerformanceOnVault(BaseModel):
    vault: Vault
    account_address: str
    total_deposited: float | None = None
    total_withdrawn: float | None = None
    current_num_shares: float | None = None
    current_value_of_shares: float | None = None
    share_price: float | None = None
    all_time_earned: float | None = None
    all_time_return: float | None = None
    volume: float | None = None
    weekly_win_rate_12w: float | None = None
    deposits: list[VaultDeposit] | None = None
    withdrawals: list[VaultWithdrawal] | None = None
    locked_amount: float | None = None
    unrealized_pnl: float | None = None


# --- Vault Operation Args ---


@dataclass
class CreateVaultArgs:
    vault_name: str
    vault_description: str
    vault_social_links: list[str] = field(default_factory=list)
    vault_share_symbol: str = ""
    vault_share_icon_uri: str | None = None
    vault_share_project_uri: str | None = None
    fee_bps: int = 0
    fee_interval_s: int = 0
    contribution_lockup_duration_s: int = 0
    initial_funding: int = 0
    accepts_contributions: bool = True
    delegate_to_creator: bool = True
    contribution_asset_type: str | None = None
    subaccount_addr: str | None = None


@dataclass
class ActivateVaultArgs:
    vault_address: str
    additional_funding: int | None = None


@dataclass
class DepositToVaultArgs:
    vault_address: str
    amount: int


@dataclass
class WithdrawFromVaultArgs:
    vault_address: str
    shares: int
